sort/timsort.go

```diff
@@ -0,0 +1,111 @@
+// Implementation of Timsort algorithm
+// Reference: https://en.wikipedia.org/wiki/Timsort
+
+package sort
+
+import (
+	"github.com/TheAlgorithms/Go/constraints"
+)
+
+const runSizeThreshold = 8
+
+// Timsort is a simple generic implementation of Timsort algorithm.
+func Timsort[T constraints.Ordered](data []T) []T {
+	runSize := calculateRunSize(len(data))
+	insertionSortRuns(data, runSize)
+	mergeRuns(data, runSize)
+	return data
+}
+
+// calculateRunSize returns a run size parameter that is further used
+// to slice the data slice.
+func calculateRunSize(dataLength int) int {
+	remainder := 0
+	for dataLength >= runSizeThreshold {
+		if dataLength%2 == 1 {
+			remainder = 1
+		}
+
+		dataLength = dataLength / 2
+	}
+
+	return dataLength + remainder
+}
+
+// insertionSortRuns runs insertion sort on all the data runs one by one.
+func insertionSortRuns[T constraints.Ordered](data []T, runSize int) {
```

Comment:
Can you not use the already implemented insertion sort and its helpers for this? I'm not sure if its possible because I haven't properly looked but I think its possible to accommodate both here.

Reply:
I originally reused the Insertion sort algorithm but I was not sure if it would be acceptable. I looked into it and I was able to reuse the merge helper function from Merge sort as well. Feel free to check it out.